File: Lista_7/lista7_ex1.c
/* Exercício 1:
Suponha que criamos uma estrutura para armazenar produtos de um supermercado:
typedef struct Produto{
    char nome[80];
    double preco;
    int quantidade;
}Produto;
Implemente duas funções, uma que devolve o vetor ordenado por preços e outra que devolve o vetor ordenado pela quantidade de itens no estoque. 
Os protótipos são:
void ordenaPreco(Produto vet[], int n);
void ordenaQuant(Produto vet[], int n);
*/
#include <stdio.h>
#include <stdlib.h>

typedef struct Produto{
    char nome[80];
    double preco;
    int quantidade;
}Produto;

void insereProd(Produto vet[], int n);
void imprimirProd(Produto vet[], int n);
void ordenaPreco(Produto vet[], int n);
void ordenaQuant(Produto vet[], int n);

int main () {
int n;
Produto vet[1000];

    printf("\nNumero de produtos a serem inseridos: ");
    scanf("%d", &n);

    insereProd(vet, n);

    printf("\n########## Lista de Produtos ##########\n");
    imprimirProd(vet, n);

    ordenaPreco(vet, n);

    printf("\n########## Lista Ordenada por Preco ##########\n");
    imprimirProd(vet, n);

    ordenaQuant(vet, n);

    printf("\n########## Lista Ordenada por Quantidade ##########\n");
    imprimirProd(vet, n);

return 0;
}

void insereProd(Produto vet[], int n) {
    for(int i=0; i < n; i++) {
        printf("\nInsira %d produto:\n", i+1);
        printf("Nome: ");
        scanf("%s", vet[i].nome);
        printf("Preco: ");
        scanf("%lf", &vet[i].preco);
        printf("Quantidade: ");
        scanf("%d", &vet[i].quantidade);
    }
}

void imprimirProd(Produto vet[], int n) {
    for(int i=0; i < n; i++) {
        printf("\nPRODUTO %d:", i+1);
        printf("\nNome: %s", vet[i].nome);
        printf("\nPreco: %.1lf", vet[i].preco);
        printf("\nQuantidade: %d\n", vet[i].quantidade);
    }
}

void ordenaPreco(Produto vet[], int n) {    
    Produto temp[1];

    for(int i=0; i < n; i++) {
        for(int j=i+1; j < n; j++) {
            if(vet[i].preco > vet[j].preco) {
                temp[0] = vet[i];
                vet[i] = vet[j];
                vet[j] = temp[0];
            }
        }
    }
}

void ordenaQuant(Produto vet[], int n) {    
    Produto temp[1];

    for(int i=0; i < n; i++) {
        for(int j=i+1; j < n; j++) {
            if(vet[i].quantidade > vet[j].quantidade) {
                temp[0] = vet[i];
                vet[i] = vet[j];
                vet[j] = temp[0];
            }
        }
    }
}

File: Lista_7/Lista7_ex1.py
class Produto:

	def __init__(self):
		self.__nome = '' #privado
		self.__preco = 0
		self.__quantidade = 0
	
	def get_nome(self):
		return self.__nome

	def get_preco(self):
		return self.__preco

	def get_quantidade(self):
		return self.__quantidade

	def cadastra(self):
		self.__nome = input("Nome: ")
		self.__preco = float(input("Preco: "))
		self.__quantidade = int(input("Quantidade: "))
	
	def mostra(self):
		print('Nome: ' + self.get_nome() +
		      '; Preco: ' + str(self.get_preco()) +
		      '; Quantidade: ' + str(self.get_quantidade()))

class Supermercado (Produto):

	def __init__(self):
		self.__lista = []
		self.__numero = 0

	def get_numero_produtos(self):
		self.__numero = int(input("Numero de produtos: "))
		return self.__numero
	
	def get_lista(self, produto):		
		produto.cadastra()
		self.__lista.append(produto)
		return self.__lista

	def ordena_produtos_preco(self):
		ordenaPreco = sorted(self.__lista, key=lambda obj: obj.get_preco())
		print("\n### Produtos ordenados pelo preco ###\n")
		for p in ordenaPreco:		
			p.mostra()
			print("----------")

	def ordena_produtos_quantidade(self):
		self.__lista.sort(key=lambda obj: obj.get_quantidade())	
		print("\n### Produtos ordenados pela quantidade ###\n")
		for q in self.__lista:		
			q.mostra()
			print("----------")

if __name__ == "__main__":
	
	s = Supermercado()

	print("### Cadastro de Produtos ###\n")
	n = s.get_numero_produtos()
	for i in range(n):
		print("----------")
		p = Produto()
		s.get_lista(p)	
	
	s.ordena_produtos_preco()
	s.ordena_produtos_quantidade()
